### ISSUE TITLE
Feature request: support `has_unit` in Lua

### ISSUE THREAD

**Issue by vgaming** [wesnoth/wesnoth#2649]
It should have a signature  "has_unit(cfg: WML): boolean"  (takes the WML tag and returns a boolean).

Current work-around:
```
local function has_unit(cfg)
	global_temp = false
	wesnoth.wml_actions["if"] {
		T.have_unit(cfg),
		T["then"] { T.lua { code = "global_temp = true" } }
	}
	return global_temp
end
```

**Comment by gfgtdf** [comment]
you can just use `#wesnoth,get_units(cfg) > 0`

**Comment by vgaming** [comment]
@gfgtdf This would load all said units as Lua entities, which is some overhead in certain scenarios. Also, `has_units` is special in that it doesn't count units killed just now inside a `die` event.
Thoughts?

**Comment by ProditorMagnus** [comment]
Exists as wesnoth.wml_conditionals.have_unit { side = "1,2"}.

**Comment by vgaming** [comment]
@ProditorMagnus thanks a lot!

**Comment by gfgtdf** [comment]
> This would load all said units as Lua entities, which is some overhead in certain scenarios

unit filters had a limit= atribute at soem point (where limit=n would abort the search after n results were found), but somwhow i cannot find it in the code right now.

> Also, has_units is special in that it doesn't count units killed just now inside a die event.

this special is baicially a check for hitpoints > 0, which also doable in a wml filter

but ye what ravanna said works too.

**Comment by CelticMinstrel** [comment]
I think adding this as a core API function is actually a good idea, so I'm gonna reopen this. However, it is indeed possible to do via `wml_conditionals` or `eval_conditional`, so it's not high priority.

**Comment by Vultraz** [comment]
What do you mean a core API function?

**Comment by CelticMinstrel** [comment]
A Lua function rather than a WML conditional action. It would probably support input other than a WML table (eg a location).

**Comment by CelticMinstrel** [comment]
To expand on that better now that I'm not on my phone:

The existing method is `wesnoth.wml_conditionals.have_unit(wml_table)` or `wesnoth.eval_conditional(T.have_unit(wml_table))` (the latter method allows you to use `[and]` and stuff). These are direct hooks into the ConditionalWML system (with the former being the place to install _new_ ConditionalWML actions). In particular, `wesnoth.wml_conditionals` is intended to be a place to add new functions, not a place to call functions from, and all such functions take only one argument, a WML table.

In order to qualify as a core API function in my opinion, a function needs to accept the parameters it knows about as separate parameters, rather than all stuffed into a Lua table. Now, in the case of `have_unit`, it just so happens that the argument is a filter, so it would look more or less the same (that is, unless we add a way to precompile filters for repeated use, or unless we follow up on my thought from the previous comment and add a few convenient shorthands for common use-cases). However, it should at least be available in the `wesnoth` namespace or in one of the new modules introduced by the Lua API reorganization, rather than only under `wml_conditionals`.

(This isn't unique to this function, it applies to literally anything that's currently only available under `wml_conditionals` or `wml_actions` or even `effects`.)

**Comment by vgaming** [comment]
In general I would agree that it's good to expose a Lua-ish API whenever possible (instead of WML-ish). In this particular case, however, I don't see any real benefits that a Lua-specific `have_unit` could provide.
The argument being a WML table doesn't hurt. Using any kind of simplification around `lua_function` doesn't really help either because you can do iteration and checking in pure Lua without using API, and it's just as or more flexible.

I'd concentrate efforts on other work TBH.

**Comment by CelticMinstrel** [comment]
Yeah, it's not a priority, but it's something that'd be nice to have someday.
